### PR TITLE
mutate: AOR schemata

### PR DESCRIPTION
**Pull request by joakim-brannstrom**

### CODE REVIEW

**Comment by joakim-brannstrom** [comment]
Screenshot_20200329_234105
Progress. It is still pretty crude but schemas are generated, executed and the results stored.

File: plugin/mutate/doc/design/notes/schemata.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+There may be schematas that isn't possible to compile.
+
+### Design
+
+The design in dextool assumes that 1) isn't possible to perfectly solve for all
+cases and that the generator that is implemented is incorrect. dextool will use
+a strategy wherein it first try to execute all applicable schematas and then go
+over to, if any are left to test, source code mutants. This should make it
+highly robust against invalid schematas but still see a speedup where schematas
+are used.
```
Comment (christoffer-nylen) [inline plugin/mutate/doc/design/notes/schemata.md:37]: Nice!👍 I think this is easy to understand.

File: plugin/mutate/doc/design/notes/schemata.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+interaction that lead to a schemata changing the behavior of the test. It would
+for example be *fatal* to the mutation testing result if a schemata affects the
+test cases so one or more of the tests start to always *fail* which would mean
+that all mutants are marked as **killed**. The result is faulty in this case.
+
+### Design
+
+The design for 2) is to introduce a sanity check for a schemata. Before a
+schemata is used the test suite is executed in order to see that it reports
+"passed" when no mutant is activated.
```
Comment (christoffer-nylen) [inline plugin/mutate/doc/design/notes/schemata.md:51]: 👍

File: plugin/mutate/doc/design/notes/schemata.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+### Design
+
+The design in dextool assumes that 1) isn't possible to perfectly solve for all
+cases and that the generator that is implemented is incorrect.
+
+dextool uses a strategy where it generates one schemata per file. It will
+result in dexotool seeing a lower speedup than is optimal from schemata but it
+makes dextool the schemata strategy robust against compilation errors. One
+compilation error only invalidates that specific schemata.
```
Comment (christoffer-nylen) [inline plugin/mutate/doc/design/notes/schemata.md:155]: Cool! 👍

**Comment by joakim-brannstrom** [comment]
I have validated the implementation locally on googletest. It all seems to work perfectly now together with some bug fixes to other things. Most notably the early stop implementation.
gtest now takes about 2-3s per mutant instead of 1-1.5 minutes.
A pretty nice speed-up :)

**Comment by joakim-brannstrom** [comment]
Now time to write some tests for this that at least covers the sunny day scenarios

**Pull request merged**